lib/winrm/http/transport.rb

```diff
@@ -103,6 +104,85 @@ def initialize(endpoint, user, pass, opts)
       end
     end
 
+
+    # NTLM/Negotiate, secure, HTTP transport
+    class HttpNegotiate < HttpTransport
+      def initialize(endpoint, user, pass, opts)
+        super(endpoint)
+        no_sspi_auth!
+        @user, @pass = user, pass
+        @ntlmcli = Net::NTLM::Client.new(user, pass)
+        @retryable = true
+        no_ssl_peer_verification! if opts[:no_ssl_peer_verification]
+      end
+
+      def send_request(message)
+        init_auth if @ntlmcli.session.nil?
+
+        original_length = message.length
+
+        emessage = @ntlmcli.session.seal_message message
+        signature = @ntlmcli.session.sign_message message
+        seal = "\x10\x00\x00\x00#{signature}#{emessage}"
+
+        hdr = {
+          "Connection" => "Keep-Alive",
+          "Content-Type" => "multipart/encrypted;protocol=\"application/HTTP-SPNEGO-session-encrypted\";boundary=\"Encrypted Boundary\""
+        }
+
+        body = <<-EOF.gsub(/^\s{10}/,"")
+          --Encrypted Boundary\r
+          Content-Type: application/HTTP-SPNEGO-session-encrypted\r
+          OriginalContent: type=application/soap+xml;charset=UTF-8;Length=#{original_length}\r
+          --Encrypted Boundary\r
+          Content-Type: application/octet-stream\r
+          #{seal}--Encrypted Boundary--\r
+        EOF
+
+        resp = @httpcli.post(@endpoint, body, hdr)
+        if resp.status == 401 && @retryable
+          @retryable = false
+          init_auth
+          send_request(message)
+        else
+          @retryable = true
+          handler = WinRM::ResponseHandler.new(winrm_decrypt(resp.body), resp.status)
+          handler.parse_to_xml()
+        end
+      end
+
+      def winrm_decrypt(str)
+        str.force_encoding('BINARY')
+        str.sub!(/^.*Content-Type: application\/octet-stream\r\n(.*)--Encrypted.*$/m, '\1')
```

Comment:
Normally I would agree but since the argument to this method is always a server response the argument should never be used again. Doing a non-destructive `#sub` on this just duplicates the string for no apparent good reason.